### ISSUE TITLE
Node Server Debug

### ISSUE THREAD

**Issue by chanpod** [Microsoft/vscode#3166]
Seems the latest update broke the node debugger.
Here's my config.

```
{
	"version": "0.1.0",
	// List of configurations. Add new configurations or edit existing ones.
	"configurations": [
		{
			// Name of configuration; appears in the launch configuration drop down menu.
			"name": "Launch server",
			// Type of configuration.
			"type": "node",
            "request": "launch",
			// Workspace relative or absolute path to the program.
			"program": "${workspaceRoot}server/server.js",
			// Automatically stop program after launch.
			"stopOnEntry": false,
			// Command line arguments passed to the program.
			"args": [],
			// Workspace relative or absolute path to the working directory of the program being debugged. Default is the current workspace.
			"cwd": "${workspaceRoot}",
			// Workspace relative or absolute path to the runtime executable to be used. Default is the runtime executable on the PATH.
			"runtimeExecutable": null,
			// Optional arguments passed to the runtime executable.
			"runtimeArgs": ["--nolazy"],
			// Environment variables passed to the program.
			"env": {
				"NODE_ENV": "development"
			},
			// Use JavaScript source maps (if they exist).
			"sourceMaps": false,
			// If JavaScript source maps are enabled, the generated code is expected in this directory.
			"outDir": null
		},
		{
			"name": "Attach",
            "request" :"attach",
			"type": "node",
			// TCP/IP address. Default is "localhost".
			"address": "localhost",
			// Port to attach to.
			"port": 5858,
			"sourceMaps": false
		}
	]
}
```

and the output I get

```
node --debug-brk=32554 --nolazy server\server.js 
Debugger listening on port 32554
{ [Error: Cannot find module '../build/Release/bson'] code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version
```

Obligatory "It worked fine before" I may be missing something since the update. It was fussing about the file paths but I updated those to what I assume to be correct (as it's no longer giving me a warning).

It launches, outputs that bit of text, and then stops. No errors. I've tried 
`"program": "${workspaceRoot}server/server.js",`
and
`"program": "${workspaceRoot}/server/server.js",`

**Comment by weinand** [comment]
@chanpod Is there a '/' missing between the `${workspaceRoot}` and `server/server.js` ?

```json
"program": "${workspaceRoot}server/server.js",
```

**Comment by chanpod** [comment]
I've tried it with and without. Same behavior. I tried deleting my launch config completely and starting with a fresh one since there were several changes. But it just kept ignoring the slash and giving the error

`"'c:\Users\cphilpot\git\bossCollectionserver\server.js"`

It does that with or without the slash in

`"program": "${workspaceRoot}/server/server",`

**Comment by weinand** [comment]
@chanpod I've just verified on Windows 10 and VS Code 0.10.8 that the '/' is definitely needed. In my test I'm using this launch config:

```json
{
	"name": "Server",
	"type": "node",
	"request": "launch",
	"program": "${workspaceRoot}/server/server.js",
	"cwd": "${workspaceRoot}",
	"stopOnEntry": true,
	"runtimeArgs": ["--nolazy"]
}
```

With it my server just starts fine with this output on the debug console:

```
node --debug-brk=27537 --nolazy server\server.js 
Debugger listening on port 27537
```

Since the output matches your output, I'm pretty sure that your server starts up fine but then has an issue with the `js-bson` module, which is not part of VS Code. You can verify this by using `"stopOnEntry": true`. Your server should stop on the first line but produces no error output.
 
The output:

```
{ [Error: Cannot find module '../build/Release/bson'] code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version
```

is not coming from VS Code. Please verify why the relative path '../build/Release/bson' is not working.

**Comment by chanpod** [comment]
MongoStore = require('connect-mongo')(session)

is throwing the error. This is a fairly common error (no idea why) and it doesn't actually prevent the server from working. But it does seem to be stopping the VS debugger from continuing. Is there a way to tell it to ignore errors?

**Comment by weinand** [comment]
@chanpod in what state is the debugger if the error occurs? Still in a debug session but stopped on a exception? A screenshot of the debugger would be helpful.

**Comment by chanpod** [comment]
http://imgur.com/2wwc547

After I step over this line, the debugger just stops and the server never finishes starting. Seems the debugger kills the entire session.

**Comment by DasithKuruppu** [comment]
Same issue here , debugger just hangs on some random line or never gets past some , using native nodejs compiler/debugger to run application works just fine but whenever running with VS code and some breakpoints it almost never does.

**Comment by tones411** [comment]
Mine works with these settings on the Mac:
...
"program": "${workspaceRoot}/server.js",
"cwd": "${workspaceRoot}",
...

**Comment by weinand** [comment]
@chanpod do you have a minimal repro that I can debug?

**Comment by nathan-walker** [comment]
@chanpod Have you tried updating your Node.js installation to the latest version?  I was experiencing the same issue on v5.4.1 and an update to v5.8.0 just fixed it for me.

**Comment by chanpod** [comment]
I'm was on Node Version 4.3
I'm upgrading to 5.8 now to see if it will work.

The config is working as expect now. Seems one of the previous updates fixed it (or it was just some strange isolated intermittent issue) But it is still silently crashing b/c of the mongoose module. Which, as stated before, works fine with a normal npm start.

Will update in a bit.

**Comment by chanpod** [comment]
It seems using Node Version 5.8 I no longer have the issue. The debugger launches and runs as expected and I'm able to successfully use break points and what not.

@weinand If you want to see the mess that is my side project,
https://github.com/chanpod/bossCollection

I didn't test Node V4.4 and Node still has that as a recommended version. I'll test it when I get home to see if I still experience the same issue.

**Comment by weinand** [comment]
@chanpod thanks for verifying your project with a newer version of node. Yes, some versions of node have obscure issues with debugging. Whenever you see a problem, it is always a good idea to try an older or a newer version.

Does the crashing b/c of the mongoose module still occurs for your github project and node 5.8? If yes, I would like to investigate it.

**Comment by chanpod** [comment]
@weinand The debugger seems to be working fine again. After updating and testing with node versions 4.4 and 5.8 and as of VS Code V. 0.10.11, I'm not able to replicate the issue anymore. I did a fresh install of all my npm modules as well so I know my temp fix is no longer in place as well. No idea what it was but it's fixed now. So, yay :+1:

**Comment by maryamsoprano** [comment]
Hi,

I tried all procedures. My express server is listening on identified port now. but after starting the web site, breakpoints does not work.
my launch.json is:
{
 "version": "0.2.0",
  "configurations": [
        {
            "name": "Launch",
            "type": "node",
            "request": "launch",
            "program": "/home/soprano/workspace/myui/server.js",
            "stopOnEntry": false,
            "args": [],
            "cwd": "${workspaceRoot}",
            "preLaunchTask": "",
            "runtimeExecutable": null,
            "sourceMaps": true,
            "runtimeArgs": [
                "--nolazy"
            ],
            "env":{
                "NODE_ENV":"development"
            },
            "externalConsole": false

        },
        {
            "name": "Attach to url with files served from ./out",
            "type": "node",
            "request": "attach",
            "outDir":"${workspaceRoot}/app",

            "port": 3001

        }
    ]
}

and this is my package.json file:

{
  "name": "myui",
  "version": "1.0.0",
  "description": "",
  "main": "server.js",
  "scripts": {
    "server": "node server.js",
 
      "start": "concurrently \"npm run tsc:w\" \"npm run server\" ",
     "tsc": "tsc",
    "tsc:w": "tsc -w -p /home/soprano/git/ui"
  },
  "author": "",
  "license": "ISC",
   "dependencies": {
    "concurrently": "^2.2.0",
    "express": "^4.14.0"
  }
   
}

what I see on the debug console is:
node --debug-brk=18135 --nolazy ../../workspace/myui/server.js 
Debugger listening on port 18135
Express server listening on port 3001!



What is wrong in my configuration?
(version of nodejs is 6.2.2 and my vscode is 1.3.1)
please help me.
thanks in advance.

**Comment by weinand** [comment]
@chanpod you have two launch configurations in your launch.json. Before we can find out what is wrong with them, we have to find out which launch config are you using.

Are you launching your server with VS Code (with the launch configuration "Launch") or are you have launching your server from the command line, e.g. via "npm" and then you attach the debugger with the launch configuration "Attach to url with files served from ./out")?

**Comment by maryamsoprano** [comment]
Thanks for your response. 
I have tried both of them. 
First, I chose lunch, and in this state, breakpoints were not triggered. And just listen to the port . And I could see localhost:port on firefox. 
Secondly, I tried attach, in this part, nothing happened. And debug did not start.

**Comment by weinand** [comment]
@maryamsoprano I've changed my original comment from above because I noticed that you are using 'launch'. So please read my comment from above again and follow my suggestions.

**Comment by maryamsoprano** [comment]
Hi, 
And thanks for your suggestion. 
My javascript is not in another folder. It is beside typescript file. 
For example we have ui/lib/contacts folder and in this folder, there are some html and js and ts files. Also some files resists in ui/lib/components(includs ts,js and html) and etc. 
In this way what should be my outdir?
And also program path is out of my project. It points the path of my nodejs(server.js). Did I assign wrong path?

**Comment by maryamsoprano** [comment]
I should add that my server.js ,just start the project as following :

var express = require('express');
var app = express();

app.use(
    "/", //the URL throught which you want to access to you static content
    express.static("/home/soprano/git/ui") //where your static content is located in your filesystem
);

app.listen(3002, function () {
  console.log('Express server listening on port 3001!');
});

in my program path, I pointed to this specific file.

**Comment by weinand** [comment]
@maryamsoprano 
The only thing you can try is to set `outDir` to the folder where the generated JavaScript lives (which in your case seems to be outside of your project `/home/soprano/workspace/myui/`).

My last comment on this issue:
Your setup seems to be very confusing (and is probably broken)...
I cannot explain how to fix it without having more details.
Sorry. 

If I should continue to help you, please do this:
- create a new issue (because this issue is unrelated to your problem and is already closed).
- explain in detail your project structure and where your other files live.
